### PR TITLE
ENHANCEMENT upgrade cropper plugin to v2.3.4

### LINKED ISSUE #13: Upgrading to Cropper.js v2.3.4
@willmorgan would there need to be a new branch if a PR for this would be submitted? I've been working through the latest tag of the plugin and many of the plugin options have changed.

I'm working to see how feasible it is to get the default options to function the same as the current version, but some things may have been dropped so in that respect there would be breaking changes.

### PR DESCRIPTION
**Pull request by muskie9**
closes #13

### CODE REVIEW

**Comment by muskie9** [comment]
@willmorgan I did convert the protected options to private static to use the config system. Let me know if you'd prefer it to not be that way and I can change it back.

**Comment by muskie9** [comment]
I'm working through the test failure but when I run that particular test locally via sake, I get a different error at [TestCase.phpL#125](https://github.com/willmorgan/silverstripe-cropperfield/blob/master/tests/CropperTest/TestCase.php#L125). I'm guessing this could be related to my local env, but just wanted to check if you've run across this in the past at all.

```
CropperTest\GenerationProcessTest::testGenerateCropped
mkdir(): No such file or directory
mkdir(assets/GenerationProcessTest)
TestCase.php:125

CropperTest\TestCase->prepareFileAssets(Image)
GenerationProcessTest.php:30
```

**Comment by willmorgan** [comment]
@muskie9 I think there's an issue with this module for newer builds of SS, its something that @dhensby and I were looking in to yesterday. I'll take a look tomorrow.

**Comment by muskie9** [comment]
Cheers, I didn't think of the newer SS version as a possible cause.

**Comment by willmorgan** [comment]
@muskie9 It would appear that there's a bit of a failure with tests that is probably related to your work - could you take a look and fix it here?

https://travis-ci.org/willmorgan/silverstripe-cropperfield/jobs/191403934#L240

Could I ask that you use just the 3.1 job in Travis for the time being? If it's failing against newer versions for the same reasons as master, don't worry about that for now.

**Comment by muskie9** [comment]
Sure thing!

**Comment by muskie9** [comment]
Green light on 3.1 test

**Comment by dhensby** [comment]
Ah, I think I know what it is. `$file->exists()` now actually checks the file exists on the file system as well as in the DB - so if the test doesn't have a real image file that matches the test record then the wrong exception will be thrown...

**Comment by dhensby** [comment]
This is the relevant code: https://github.com/willmorgan/silverstripe-cropperfield/blob/e5ad57df5c1bd6826b40a6dc18465cb215306de1/code/CropperField.php#L228-L230

This fixture needs the full path: https://github.com/willmorgan/silverstripe-cropperfield/blob/e5ad57df5c1bd6826b40a6dc18465cb215306de1/tests/Fixture/GenerationProcessTest.yml#L6

**Comment by muskie9** [comment]
Is there a trick to getting the namespaced tests to run via sake? `sake dev/tests/GenerationProcessTest` doesn't seem to do anything.

**Comment by muskie9** [comment]
Trying a full path to the file to see if it fixes additional tests. If it works I'll squash.

**Comment by dhensby** [comment]
no dice - there shouldn't be any issue with namespaced tests :/

**Comment by dhensby** [comment]
oh - dont use sake

`vendon/bin/phpunit cropperfield/tests`

**Comment by muskie9** [comment]
That seems to react quicker, but getting a class not found... I presume that class should be picked up automagically since it's in the same namespace so no `use` should be needed:

```
$ vendor/bin/phpunit cropperfield/tests
PHP Fatal error:  Class 'CropperTest\TestCase' not found in /var/www/public/silverstripe.demo/cropperfield/tests/CropperTest/CropperFieldTest.php on line 10
PHP Stack trace:
PHP   1. {main}() /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/composer/bin/phpunit:0
PHP   2. PHPUnit_TextUI_Command::main($exit = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/composer/bin/phpunit:63
PHP   3. PHPUnit_TextUI_Command->run($argv = *uninitialized*, $exit = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/TextUI/Command.php:129
PHP   4. PHPUnit_Runner_BaseTestRunner->getTest($suiteClassName = *uninitialized*, $suiteClassFile = *uninitialized*, $suffixes = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/TextUI/Command.php:150
PHP   5. PHPUnit_Framework_TestSuite->addTestFiles($filenames = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Runner/BaseTestRunner.php:96
PHP   6. PHPUnit_Framework_TestSuite->addTestFile($filename = *uninitialized*, $phptOptions = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Framework/TestSuite.php:416
PHP   7. PHPUnit_Util_Fileloader::checkAndLoad($filename = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Framework/TestSuite.php:355
PHP   8. PHPUnit_Util_Fileloader::load($filename = *uninitialized*) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Util/Fileloader.php:76
PHP   9. include_once() /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Util/Fileloader.php:92

Fatal error: Class 'CropperTest\TestCase' not found in /var/www/public/silverstripe.demo/cropperfield/tests/CropperTest/CropperFieldTest.php on line 10

Call Stack:
    0.0039     226504   1. {main}() /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/composer/bin/phpunit:0
    0.0182     561640   2. PHPUnit_TextUI_Command::main(???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/composer/bin/phpunit:63
    0.0182     562264   3. PHPUnit_TextUI_Command->run(???, ???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/TextUI/Command.php:129
    0.0247     808640   4. PHPUnit_Runner_BaseTestRunner->getTest(???, ???, ???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/TextUI/Command.php:150
    0.0362    1005720   5. PHPUnit_Framework_TestSuite->addTestFiles(???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Runner/BaseTestRunner.php:96
    0.0362    1006736   6. PHPUnit_Framework_TestSuite->addTestFile(???, ???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Framework/TestSuite.php:416
    0.0381    1093944   7. PHPUnit_Util_Fileloader::checkAndLoad(???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Framework/TestSuite.php:355
    0.0384    1094152   8. PHPUnit_Util_Fileloader::load(???) /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Util/Fileloader.php:76
    0.0390    1125016   9. include_once('/var/www/public/silverstripe.demo/cropperfield/tests/CropperTest/CropperFieldTest.php') /var/www/public/silverstripe.demo/vendor/phpunit/phpunit/PHPUnit/Util/Fileloader.php:92
```

**Comment by muskie9** [comment]
Disregard my previous comment, realized I didn't have a `phpunit.xml` file

**Comment by muskie9** [comment]
Green lights on all but (job #60.7)[https://travis-ci.org/willmorgan/silverstripe-cropperfield/jobs/192561230]. Weird error, wondering if the test just needs to be restarted. If this checks out I'll squash.

**Comment by dhensby** [comment]
I've restarted the test

File: code/CropperField.php [hunk 1]
```diff
@@ -26,39 +26,123 @@
...
-	    'crop_max_height' => null,
-	    /**
-	     * The maximum dimensions a generated image can be, in px.
-	     * If the crop is above this, then it will be downscaled according to the declared aspect ratio, or
-	     * the implied aspect ratio if one is not specified.
-	     */
-	    'generated_max_width' => 512,
-	    'generated_max_height' => 512,
-	);
+    private static $cropper_options = [
```
Comment (dhensby) [inline code/CropperField.php]: We support PHP 5.3 so you can't use short array syntax - that's why the test is failing.

File: code/CropperField.php [hunk 2]
```diff
@@ -108,8 +198,8 @@ public function setForm($form) {
 	 * @param array $options
 	 * @return $this
 	 */
-	public function setOptions(array $options) {
-		$defaults = static::$default_options;
+	public function setOptions(array $options = []) {
```
Comment (dhensby) [inline code/CropperField.php]: short array syntax

**Comment by willmorgan** [comment]
Thanks for looking over this @dhensby!

**Comment by dhensby** [comment]
Right, I've pushed up the fix and squashed for you :)

**Comment by dhensby** [comment]
and we are green - anyone want to actually check it out manually and test?

**Comment by muskie9** [comment]
Thanks for catching the array syntax @dhensby. We've been out of 5.3 for a while so I've gotten used to the shorter syntax.

I'll run through some more manual tests, but as I was wrapping up the unit tests I was manually testing as well and seemed to get expected results.

**Comment by muskie9** [comment]
After setting up both a front-end and cms form, both are working as expected with various field option settings. I think this looks good from my end.

**Comment by willmorgan** [comment]
@muskie9 Released as 2.0.0: https://github.com/willmorgan/silverstripe-cropperfield/tree/2.0.0

**Comment by muskie9** [comment]
Cheers!

**Pull request merged**